doc/SVGATextMode.man
.TH SVGATextMode 8 "October 1998" "Version 1.9"

.SH NAME
SVGATextMode \- Textmode manipulation/enhancement tool

.SH SYNOPSIS
\fBSVGATextMode\fP [ -ndhrfcvsam ] [ -t \fIConfigFile\fP ]
[ \fITextmodeLabel\fP ]

.SH DESCRIPTION
\fBSVGATextMode\fP provides a means to seriously enhance the looks of your
Linux text consoles, by re-programming the (S)VGA hardware. It uses a
configuration file similar to the one the XFree86 X-windows server uses
(Xconfig or XF86Config) to set up better looking textmodes. (=higher
resolution, larger font size, higher display refresh...).

It works independently of what text modes are allowed from the BIOS. In
theory, any text mode size is possible, as long as your VGA card is up to
the job. Whether it is or not, depends greatly on the chipset.

SVGATextMode can resize the console screen on the fly, without rebooting, to
any other size. The configuration file holds all the parameters needed for
the new mode.

It needs kernel version \fB1.1.54\fP or newer if screen resizing is to be
possible.

.SS Supported chipsets
.TP
.B VGA
Generic VGA chips. This can also be used for unsupported VGA chips, but with
very limited possibilities. Most portable VGA computers (Compaq LTE, ...)
should work with this also: VGA LCD's can't use higher dot-clocks anyway.

It is important to start from a standard VGA text mode (80x25...80x50) when
using this "chipset" for unsupported chips. Otherwise the clock will not be
programmed correctly.

.TP
.B ET6000

.TP
.B ET4000

.TP
.B ET3000

.TP
.B S3
any S3-based card, including those from Diamond, Number 9 and SPEA/Video7.

.TP
.B CIRRUS
Cirrus Logic chipsets. This works with CL-GD542x, CL-GD543x ("Alpine") and
CL-GD546x ("Laguna") chipsets.

.TP
.B TVGA8900, TVGA9000
The oldest Trident cards. NOT for the Trident accelerated cards like the
9440 etc.

.TP
.B TGUI
All accelerated Trident cards from TGUI9320LCD and up.

.TP
.B PVGA1, WDC90C0X, WDC90C1X, WDC90C2X, WDC90C3X
All (?) Western Digital cards, from the plain old first Paradise card
(PVGA1) to the more recent WDC90C33 accelerator (WDC90C3X).

.TP
.B ATI
All ATI cards BEFORE the MACH32.

.TP
.B ATIMACH32
ATI MACH32 and MACH64. Many MACH64 cards use special RAMDACs and clockchips
that are not yet supported by SVGATextMode. Those will not work.

.TP
.B ATIMACH64
Only MACH64 boards with internal clockchips are supported.

.TP
.B VIDEO7
Headland Technologies based Video 7 boards only. Older V7 boards use C&T or
Cirrus chips. Newer V7/SPEA cards use S3.

.TP
.B ALI, AL2101
Avance Logic chipsets. It's not sure whether this will work on ALL Avance
Logic cards.

.TP
.B OTI67, OTI77, OTI78
Oak Technology chipsets.

.TP
.B SIS
SiS chipsets. UNTESTED!

.TP
.B RealTek
RealTek chipsets. UNTESTED!

.TP
.B ARK
ARK1000 and ARK2000 chipsets. UNTESTED!

.TP
.B NCR77C22E, NCR77C32
NCR chipsets. The NCR77C21 and NCR77C22 (without "E" suffix) will not
benefit from this. They should work just as well with the generic VGA
driver. UNTESTED!

.TP
.B GVGA
Genoa 6000 series cards. The 5000 and 7000 series are based on resp. ET3000
and ET4000 chips. Use the ET4000 driver for those. UNTESTED!

.TP
.B MX
MX graphics chips. MX86000 and MX86010 chips should work. UNTESTED!

.TP
.B MATROX
Matrox Millennium and Mystique are supported.

.TP
.B NEOMAGIC
Those NeoMagic laptop controllers

.TP
.B RIVA128
all RIVA128 derivatives (including RIVA128ZX and TNT).

.PP
SVGATextMode needs a configuration file with a similar syntax as
\fBXconfig\fP or \fBXF86Config\fP, the configuration file for XFree86, the
X-Windows server. The default config file is \fB/etc/TextConfig\fP.

See the
.B TextConfig(5)
manual file for details on its syntax.

When running SVGATextMode with a new textmode, it will output a single line,
describing what the new mode is, and what are the screen refresh rates.

.SS Example output:
Chipset = 'S3', Textmode clock = 75.00 MHz, 116x48 chars, CharCell = 9x16.
Refresh = 55.93kHz/69.9Hz.

.SS OPTIONS
The only required option is the \fBTextmodeLabel\fP, which tells
SVGATextMode to which mode it must switch. A mode description line with the
same name must be present in the config file. The parameters on that line
will be used to program the new text mode.

.TP
.B -n
don't program the new mode. This option will scan the config file for the
requested mode, parse all parameters, and report what the new mode will look
like. But it doesn't actually change anything. This is useful for seeing if
a mode is correctly entered in the config file, and if your monitor will be
able to handle it.

.TP
.B -d
debugging mode. SVGATextMode will output lots of debugging information about
all the things it attempts to do. This is mostly for "internal purposes"
only, but it could help you discover why SVGATextMode doesn't react as you
expect it to. This could also be useful when reporting a problem.

A second "d" option will enable extended debugging, showing you all the data
that was parsed from the config file, instead of just the most important
things. More specifically, this will add all the mode lines and the font
selection list to the debug output. 

.TP
.B -h
prints out a short help and the version number. 

.TP
.B -r
don't run the \fBResetProg\fP.

When a ResetProg is defined (and thus enabled) from the config file,
SVGATextMode, without "-r", will attempt to run it whenever it has changed
the screen size.

The reset program is normally used to signal or restart a program that
depends on the current screen size, and that doesn't react to the
\fBSIGWINCH\fP signal the kernel sends it upon a screen resize. Typical
examples are gpm and selection; two very handy textmode-mouse handlers.

Resizing the screen from 80x25 to e.g. 116x48 without telling gpm or
selection about it, will cause the mouse cursor to get stuck in the upper
left 80x25 characters of the screen.

The "-r" option is useful when SVGATextMode is started from the system
initialisation files (/etc/rc.d/...), because many of the programs that need
to be reset through the \fBResetProg\fP have not yet been started, so trying
to send them signals, or trying to kill them, would fail. And trying to
restart them would cause two copies of the same program running when the
rest of the rc-files start them up.

.TP
.B -f
don't run the \fBFontProg.\fP When the font loader is enabled through the
\fBLoadFont\fP option in the config file, SVGATextMode will not run it, even
when the screen is resized. When this option is not enabled and the font
loader is enabled, SVGATextMode will run an external program to load a
suitable font.

.TP
.B -c
don't program the pixel clock. More of a debugging option. If you suspect
that SVGATextMode fails to program your VGA clock properly, then this option
programs everything else \fBexcept\fP the pixel clock, which is left at its
old value. Some laptops don't react to well when the pixel clock is
reprogrammed, so this option can help there, too.

.TP
.B -v
don't validate H/V frequencies. Normally SVGATextMode will check the
horizontal frequency and the vertical refresh of the new mode against the
ranges specified in the config file, and refuse to program a mode that
doesn't fall within those limits. This option turns off that checking, and
will program ANY mode, even if the config file says your monitor cannot
handle them.

.TP
.B -s
scan mode. SVGATextMode will scan the entire config file, and will dump a
listing of all valid modes for your configuration, taking the limits for
vertical refresh, horizontal scan rate, maximum text mode pixel clock and
availability of the required clock frequency (if you don't have a
programmable clock chip) for your VGA card / monitor configuration into
account. When combined with the \fB-v\fP option, it will dump all modes,
whether they are allowed or not.

This is particularly useful to create input to a script or program that lets
you (for example) select a text mode from a menu.

.TP
.B -m
allow trying to resize screen via a 1x1 screen. This option is only useful
when you get an \fBout of memory\fP error when resizing the screen. Normally
SVGATextMode will just exit and tell you there was not enough memory. The
best thing to do then is to free some memory. If that is impossible for some
reason, this option will first resize the screen to a 1x1 size to free the
memory already taken up by all the consoles, and then resize to the desired
size.

\fBWARNING\fP: see the \fBBUGS\fP section for more information on the
potential dangers of this method. Do not use this option unless it is
absolutely necessary.

.TP
.B -a
always do a full resize. SVGATextMode will normally check with the old
screen size to see if the new mode actually resizes the screen (it could
just \fBenhance\fP the screen by moving to a larger font, but remain at the
same width/height). It does that via the tty settings (those you get when
typing "stty -a"). If it detects that the new mode is a different size than
the old one, it will run all sorts of extra code to make various system
components know about the resizing. But if it thinks no actual resizing has
been done, it won't do that.

Under certain circumstances (see the \fBBUGS\fP section), it is desirable to
resize the screen, even if the tty settings say the screen already has that
size. This option will force the resizing code to "fake" a resize, and run
all necessary code (set the tty settings again, tell the kernel about the
resizing, and run the optional \fBResetProg\fP).

.TP
.B -k
Don't tell the kernel if the screen size has changed. If the horizontal
screen size really has changed, the kernel will not know about it and render
text in completely wrong places. Some people seem to want to use this to lie
to the kernel about the amount of text lines on screen, and e.g. use the
lower line for a status display (since the kernel doesn't know it's there,
it doesn't render text in it).

As a side effect, none of the programs running on the console will be sent a
SIGWINCH from the kernel, so they too will not know about the change. The
ResetProg is still ran (unless when prohibited by the "-r" option of
course), so it could be used to send explicit resize messages to console
applications that otherwise relied on the kernel SIGWINCH signal.

.TP
.B -x
always run SVGATextMode, even if you are not running it from a normal VGA
textmode console. This disables the "dumb-user protection" that was added to
avoid people from messing up their XFree86 display when running SVGATextMode
from there. This option is only useful if you know what you're doing. It
could be used to try to restore a broken textmode using after the
Xserver crashed (in combination with the "-o" option).

.TP
\fB-t\fP \fIConfigFile\fP
This option tells SVGATextMode to use a different configuration file than
the default one (\fB/etc/TextConfig\fP).

.TP
.B -o
Force all standard VGA registers to a known textmode state. This is quite
useful when some VGA-aware program (X-server, svgalib, ...) crashes or gets
killed and leaves the screen in graphics mode.

There's only one problem with this. Most of these types of crashes leave the
keyboard in raw (scan-code) mode, and thus you won't be able to type the
command to restore text mode... You will have to do a "kbd_mode -a" from a
remote terminal first. Unless you add "kbd_mode -a" to the file defined in
the "ResetProg" label in the TextConfig file, and have "stm -o" under a
mousebutton combination with the GPM mouse package.

Another fun-stopper is the extended VGA registers, which this option does
_not_ reset, because they are chipset-dependent. This means that the "-o"
option might not help at all, depending on how sophisticated your card is.
SVGATextMode knows about a few cards and how to kick them back into "normal"
mode, but not all (due to lack of documentation and cards to test -- hint,
hint).

.SS Example
A small example of how SVGATextMode goes to work: suppose you started up in
80x25 text mode, and wanted to change to a more respectable 116x48 mode with

HISTORY
?? ??? ???? - version 1.9

- (1.8.1) remove "strip" option for "install" target (causes errors on older
  systems)
- fix sync polarity bug in Ti3026 Sync-On-Green code (Giacomo COMES)
- update vgaset's licensing to GPL (Igor Grobman/Debian, Greg Lehey)
- fix the bogus H/V VGA register upper limits. Especially vertically the
  limit was 4096, but should have been 1023 instead, which caused a lot of
  modes that appeared to stretch far below the bottom of the screen). Also
  fix all modelines that violated this constraint in the TextConfig file.
  (with help from Udo van den Heuvel)
- (1.8.2) some changes to STM_reset script (Suggested by Igor Grobman, Debian)
- DOS compile fixes and enhancements by Molnar Laszlo. New version of cwsdmpi
  included.
- don't bail out when ResetProg or FontProg doesn't exist (suggested by Igor
  Grobman, Debian)
- some FAQ additions
- fix S3 Sync-on-Green code for Ti3026. (Timo Giesselmann)
- don't allow ANYONE to run STM from a non-virtual terminal, add "-x" option
  to override this (suggested by Yann Dirson, Debian)
- set default fontpath to empty string, because "setfont" knows it as well
  (suggested by Yann Dirson, Debian)
- add small text explaining how to install STM using the SYSV init method in
  contrib/sysvinit (Michael De La Rue)
- add 40x25 mode (Otto J. Makela, Riku Saikkonen)
- (1.8.3) add Neomagic support (Pavel Machek)
- add "-k" option to disable software resize (hardware is resized) (Pavel
  Machek)
- (1.8.4) add MACH64 Rage II support (should now support all MACH64 boards
  with internal programmable clock)
- fix ET6000 MClk setting (use safer method recommended by Tseng)
- LOTS of work to program MACH64 DSP properly for textmode clocks. Thanks to
  Marc La France (of XFree86 fame) for inputs on this. 
- (1.8.5) integrate RIVA128 support from Martijn van Oosterhout.
- (1.8.6) fix bug in Neomagic code (option topleft didn't work)
- fix segfault in ET6000 "special" code
- force correct clockchip for some chips if not defined
- fix RIVA128 clock code (Martijn van Oosterhout)
- LOTS of extra free fonts (Frank Klemm)
- (1.8.7) Yikes... the RIVA code used a FIXED (#define) PCI base address.
  use PCI code now.


02 DEC 1997 - version 1.8

- add Matrox Millennium II support (jurriaan <thunder7@xs4all.nl>)


23 NOV 1997 - version 1.7

- (1.6.1) use /dev/tty0 instead of /dev/console (finally!)
- fix two minor bugs (missing break in vga_prg.c, new include for libc6)
  (Molnar Laszlo)
- added some more cards to PCI detected devices (avoids "PCI unknown device"
  message). Some cleanups in the PCI debug messages (was too verbose).
- Andree Borrmann's DOS/DJGPP patches (now Mystique should work in DOS)
- documentation updates (SIGWINCH, GPM, SDRAM, how to raise textmode max
  clock)
- (1.6.2) protect users from running STM from inside X or remotely (only
  root can do that now). (Benson L Chow)
- add "hshift" parameter to modelines (Massimiliano Ghilardi)
- (1.6.3) add support for Mystique rev 3 (Martin Scherbaum, S.u.S.e.)
- change setpalette-to-kernel color ordering from straight VGA to ANSI. This
  should make setpalette store the correct colors now. (Massimiliano
  Ghilardi)
- some Alpha compilation fixes (Antti Barck)
- (1.6.4) add Cirrus Laguna clockchip support (CLGD546x)
- add ISO_FONT9 option (Tuukka Toivonen)
- partial fix for IBM RGB RAMDAC problem with standard VGA clocks
- change "PROBLEMS" file into a FAQ. Maybe people will read it then...
- some DJGPP/DOS compilation fixes (missing #includes).


28 JUN 1997 - version 1.6

- (1.5.1) fixed bug in 9-pixel char mode (Stuart Lamble)
- fixed SIGSEGV in Cirrus clock code (Stanislav Voronyi)
- fixed bug in ET4000 W32i special_svga.c (linear address map comparator)
- (1.5.2) one more Tseng register is reset in special_svga.c (CRTC 0x33)
- fix bug in setVGAreg for 'io' registers (missing "break")
- add MATROX Mystique support (Trey Boudreau)
- (1.5.3) add a few new fonts (Massimiliano Ghilardi/Michele Maltoni)
- add DOS screen resizing patches and font loading from Andree Borrmann
- add video modelines text in contrib (PAL/NTSC)
- better DOS manual file formatting by jkekoni@aurora.tky.hut.fi
- MClk max limit is now 100 MHz instead of 70 MHz.
- the LSM file in version 1.5 was 1 year off :-)


15 APR 1997 - version 1.5

- (1.4.1) added "-k" option to setpalette (as requested)
- explicitly set a few more registers on ET4000/ET6000 to get better
  textmode restoration after X-server crash. Works much better now.
- SVGATextMode.c: added "-o" option, for "set80" functionality.
  (setting all std. VGA regs by force).
- setVGAreg.c: added "-c" and "-s" options for defining a set-mask or a
  clear-mask instead of an absolute value.
- DOS programs now use DJGPP cross-compiler from Linux.
  (doesn't work yet...)
- setVGAreg: "idx" and "io" 'registers' added for generic IO access.
- updated some docs (there was some obsolete stuff)
- fix bug in resize_specified_vts() (endless loop resizing first specified
  VT)
- only vga_prg.h includes asm/io.h now
- (1.4.2) fixed dos-gcc stuff in makefile: now "make dos" makes DOS
  executables out of the box (finally -- no more rebooting or dosemu)
- some stuff added to allow grabmode (not the others!) to compile under 
  the Cygnus GCC for Windows. Clock probing does not work yet.
- user_time() code added from Harald Koenig.
- (1.4.3) added Trident TGUI support contributed by Massimiliano Ghilardi.
- Integration of XFree 3.2 clock code (XFree/common_hw).
  THIS MAY CAUSE SOME SERIOUS NEW BUGS!
- (1.4.4) Mach64 support -- not finished yet. Not at all...
- 9-pixel wide characters now correctly display "box" characters (Stuart
  Lamble)
- added PCI bus code from XFREE/common_hw/xf86_PCI.* This will be needed for
  some of those PCI-only cards (Matrox Millennium et al).
- (1.4.6) Matrox Millennium added (not working yet)
- (1.4.7) ET4000 textmode restore enhancements
- Matrox Millennium code fixed -- ChipSet MATROX / ClockChip ti3026 (finished

SVGATextMode.c
/*  SVGATextMode -- An SVGA textmode manipulation/enhancement tool
 *
 *  Copyright (C) 1995-1998  Koen Gadeyne
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */


/***
 *** SVGATextMode.c: the great big wolf
 ***
 *** Edited on 132x64 screen. Hence the LONG lines. For best viewing conditions, use this program to make itself more readable :-)
 ***/


#include "misc.h"
#include "ttyresize.h"       /* must be first ! */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <math.h>

#ifndef DOS
#  include <linux/fs.h>
#  include <sys/ioctl.h>
#  include <sys/kd.h>
#  include <ctype.h>
#endif

/*
 * this is not very clean, but it throws out all unix tty resizing stuff, which also makes it compile clean in DOS
 * However, DOS DOES support resizing. Don't let that confuse you. This is a hack for easy compiling.
 */
#ifdef DOS
#  define NO_RESIZE
#endif


#define CHIPSETREC 1  /* included chipset definition struct */
#include "chipset.h"
#include "vga_prg.h" 
#include "unlock_svga.h" 
#include "special_svga.h" 
#include "setclock.h"
#include "validate.h"
#include "file_ops.h"
#include "messages.h"
#include "kversion.h"
#include "run_extprog.h"
#include "dump_cfgdata.h"
#include "XFREE/xfree_compat.h"

/*
 * this adds all the global variables that hold the data from the config file.
 */
#include "cfg_data.h"

/*
 * other global variables
 */

char *CommandName;
#ifndef DOS
  char *ConfigFile=CONFIGFILE; 
  #define DIR_SEP '/'
  #define DIR_SEP_S "/"
#else
  char *ConfigFile="STMText.cfg";
  #define DIR_SEP '\\'
  #define DIR_SEP_S "\\"
#endif
bool debug_messages=FALSE;
int debug_level=0;

#define LAST_CHAR(s) (((char *)s)[strlen(s) - 1])

/*
 * yacc parser functions
 */

extern int yyparse(void);
extern FILE *yyin;
#define param_file yyin

void usage()
{
     PMESSAGE(("version %s. (c) 1995-1998 Koen Gadeyne.\n"
     "  Usage: %s [options] [textmodelabel]\n"
     "  Options: -n  Don't program VGA hardware\n"
     "               (parse config file and report mode parameters)\n"
     "           -d  print debugging information\n"
     "           -h  print usage information\n"
     "           -r  don't run ResetProg\n"
     "           -f  don't run FontProg\n"
     "           -c  don't change pixel clock\n"
     "           -v  don't validate H/V frequencies with limits in config file\n"
     "           -s  scan and print out all valid text modes from the config file\n"
     "           -a  always do a full resize, even if changing to same size screen\n"
     "           -k  resize the screen, but don't tell the kernel about it.\n"
     "               side effect: no SIGWINCH will be sent to all programs\n"
     "               running on the console. The ResetProg is still called\n"
     "           -x  run even when not in textmode (can screw up your console!)\n"
     "           -m  allow 1x1 screen to avoid `VT_RESIZE: out of memory' error\n"
     "               (relatively dangerous - read SVGATextMode(8) manual first!)\n"
     "           -t <ConfigFile>\n"
     "               Use <ConfigFile> instead of the default (%s)\n"
     "           -o  force all standard VGA registers to a known state. See docs!\n"
     "               (useful to restore textmode when some graphics-program crashed)\n"
     "  Textmodelabel: an existing label from the config file\n"
     "    (optional when `DefaultMode' defined in config file)\n",
     VERSION, CommandName, ConfigFile));
}

/****************************************************************************************************************************/


int main (int argc, char* argv[])
{
  char *req_label=NULL;

  bool program_hardware=TRUE;
  bool scanmodes=FALSE;
  bool validate=TRUE;
  bool alwaysresize=FALSE;
  bool run_from_anywhere=FALSE;
  bool resize1x1=FALSE;
  bool force_all_regs=FALSE;
  bool no_kernel_resize=FALSE;
  int c;


  bool sresize=TRUE;         /* will be set when screen has been resized and vice-versa */

  bool run_resetprog=TRUE, run_fontprog=TRUE, run_pixclock=TRUE;

  t_mode *curr_textmode;

#ifndef DOS
  char *cptr; /* bc3c */

  const char *cdev = ConsoleDevice("");
  size_t clen = strlen(cdev);
#endif

 /*
  * command-line argument parsing
  */

  CommandName = argv[0];

  while ((c = getopt (argc, argv, "ndhrfcsakxmvot:")) != EOF)
    switch (c)
    {
      case 'n': program_hardware=FALSE;
                break;
      case 'd': debug_level++;
                debug_messages=TRUE;
                break;
      case 'h': usage();
                exit(0);
                break;
      case 'r': run_resetprog=FALSE;
                break;
      case 'f': run_fontprog=FALSE;
                break;
      case 'c': run_pixclock=FALSE;
                break;
      case 's': scanmodes=TRUE;
                break;
      case 'm': resize1x1=TRUE;
                break;
      case 'v': validate=FALSE;
                break;
      case 'a': alwaysresize=TRUE;
                break;
      case 'k': no_kernel_resize=TRUE;
                break;
      case 'x': run_from_anywhere=TRUE;
                break;
      case 't': ConfigFile=safe_strdup(optarg);
                break;
      case 'o': force_all_regs=TRUE;
                break;
      case '?': usage();
                PERROR(("Bad option '-%c'\n",(char)optopt));
                exit(-1);
                break;
      default: PERROR(("getopt returned unknown token '%c'.\n",c));
    }

  PVERSION;
  PDEBUG(("Debug level: %d\n", debug_level));
    

 /*
  * bc3c: don't allow running stm in some cases (user protection :-)
  */
#ifndef DOS
  if (!run_from_anywhere)
  {  
    if ((cptr=ttyname(STDIN_FILENO))) {
      PDEBUG(("running from %s\n", cptr));
      if (strcmp(cptr, "/dev/console"))
        if ( !((!strncmp(cptr, cdev, clen)) && (isdigit(cptr[clen]))) )
      {
        PERROR(("You can only run SVGATextMode from a virtual terminal "
                "(=not in X or remotely), unless you use the '-x' option.\n"));
      }
    } else {
      PERROR(("stdin is not a tty\n"));
    }
  }
#endif

 /*
  * open parameter file, parse the sucker, close up again.
  */
  param_file = open_param_file(ConfigFile);
  PDEBUG(("Parsing Config file...\n"));
  while (!feof(yyin)) { yyparse(); }
  fclose(param_file);

 /*
  * Do some sanity checks
  */

  sanitize_cfgfile_data();
   
 /*
  * show all parsed data
  */
  
  if (debug_level>0) dump_cfgdata(debug_level);

  /*
   * switch between "scan" mode or normal mode
   */
  
  if (scanmodes)
  {
    scan_valid_modes(validate);
    exit(0);
  }
  else   /* normal program-a-new-textmode mode */
  {
   /*
    * look for a text mode label. If none, look for a "defaultmode" in the config file
    */
    req_label=NULL;

    /* if requested mode is on command line: use that one */
    if (optind < argc) req_label = argv[optind];
    
    /* defaultmode in config file is default when not overridden by command line */
    else if (defaultmode)
      {
        req_label = defaultmode;
        PDEBUG(("No mode label on command line: using '%s' as default mode\n", req_label));
      }

    /* none of the above: eject! eject! */
    else
      {
        usage();
        PERROR(("No textmode label on commandline, and no 'DefaultMode' in config file.\n"));
      }
  }


/* find last occurence of requested text mode line: This will allow
 * user-patched mode lines at end of TextConfig file to get preference over the
 * ones above it, which normally are the default ones (as suggested by Kenneth
 * Albanowski). Due to the way the file is parsed, the last lines are first in
 * the config file data structure, so we match the first occurence in the data base.
 */

  curr_textmode = text_mode_list;
  while (curr_textmode)
  {
    if (!strcasecmp(curr_textmode->name, req_label)) break;
    curr_textmode = curr_textmode->next;
  }
  
  if (!curr_textmode) PERROR(("Text Mode `%s' could not be found in the config file.\n", req_label));

  /* requested mode found in data base. show some information */
  PDEBUG(("Found requested text mode in config file:\n"));
  dump_mode(curr_textmode);
  
  /* scale the cursor to fit the chosen font size */
  cursor_start = (cursor_start * curr_textmode->FontHeight) / 32;
  cursor_end = (cursor_end * curr_textmode->FontHeight) / 32;
  PDEBUG(("Cursor start-end = %d-%d\n", cursor_start, cursor_end));

  /* scale underline position to fit current font size */
  if (underline_pos>=0) underline_pos = (underline_pos * curr_textmode->FontHeight) / 32;
  PDEBUG(("Underline pos. will be: %d\n", underline_pos));
 

 /*
  * show the user what mode will be programmed, and also check if it is allowed.
  */
  
  check_and_show_mode(curr_textmode, validate);

 /*
  * start changing some things now.
  */

  if (program_hardware)
  {
     sresize = check_if_resize(curr_textmode->cols, curr_textmode->rows);

     /* FIRST check for IO permissions, to avoid first resizing the screen, and then seeing that we can't
        write to the VGA regs */
         
     get_VGA_io_perm(chipset);
     
     /* sync disks if requested. Is there any way to do this and be SURE _all_ data has been flushed? */
     if (OFLG_ISSET(OPT_SYNC))
     {
       PMESSAGE(("Syncing disks...\n"));
       sync(); sleep(2); /* same as in "reboot" command. Hopefully enough */
     }
     
#ifndef DOS
#  ifndef NO_RESIZE
   
     if ((sresize || alwaysresize) && (!no_kernel_resize))
     {
      /*
       * first see if current kernel version supports resizing.
       */
       
       if (!check_kernel_version(1,1,54, "Virtual Terminal resizing"))
         PERROR(("Screen resizing not allowed (kernel version must be >= 1.1.54). Use a non-resizing text mode, or upgrade your kernel.\n"));

      /*
       * Resize the screen. Still needs LOTS more error checking to avoid dropping out in the middle, leaving
       * the user with a garbled screen.
       *
       * sresize will be TRUE when resizing tty's should be forced (due to the 2nd attempt do_VT_RESIZE will do
       * when not enough memory is free).
       *
       */

        /*
         * ALWAYS do a VT_RESIZE, even if we already did a VT_RESIZEX on a 1.3.3 or higher kernel, 
         * until those kernel programmers make this unambiguous
         */
      
       if (do_VT_RESIZE(curr_textmode->cols, curr_textmode->rows, resize1x1)) sresize=TRUE;
         
       if (check_kernel_version(1,3,3, "VT_RESIZEX"))
         {
           /*
            * VDisplay must de divided by 2 for DoubleScan modes,
            * or VT_RESIZEX will fail -- until someone fixes the kernel
            * so it understands about doublescan modes.
            */
           if (do_VT_RESIZEX(curr_textmode->cols,
                             curr_textmode->rows,
                             curr_textmode->VDisplay / (MOFLG_ISSET(curr_textmode, ATTR_DOUBLESCAN) ? 2 : 1),
                             curr_textmode->FontHeight,
                             curr_textmode->HDisplay/8*curr_textmode->FontWidth,
                             curr_textmode->FontWidth, resize1x1)) sresize=TRUE;
         }
         
      /*
       * resize terminals. If specified in "terminals" line, do just those.
       * If not specified, find out which ones are active, and resize those.
       * This is obsoleted by kernel v1.3.3 and up (they do it themselves upon
       * a call to VT_RESIZE)
       */
          
       if (!check_kernel_version(1,3,3, "Automatic TTY resizing (SIGWINCH)"))
       {
         if (p_terminals) resize_specified_vts(curr_textmode->cols, curr_textmode->rows);
           else resize_active_vts(curr_textmode->cols, curr_textmode->rows);
       }
     }

#  else 
     /*
      * no resizing support: do NOT allow resizing. Check if some stubborn moosehead tries anyway.
      */
      
      if (sresize) PERROR(("Resizing is not allowed (NO_RESIZE defined during compilation).\n"));
#  endif
#else   /* DOS */
      if (sresize) resize_DOS(curr_textmode->cols, curr_textmode->rows, curr_textmode->FontHeight);
#endif   

   /*
    * now get to the REAL hardware stuff !
    */

    Set_Textmode(force_all_regs);  /* just in case some jerk set us in graphics mode */

    unlock(chipset);

    special(chipset); /* change chipset-specific things, if needed */

    if (run_pixclock)
      do_clock(chipset, curr_textmode->pixelClock, curr_textmode->FontWidth);
    else PDEBUG(("Clock will NOT be programmed! (due to command line switch '-c') !\n"));
    
    Set_MAX_SCANLINE (curr_textmode->FontHeight);
    
    set_V_timings(curr_textmode->VDisplay, curr_textmode->VSyncStart, curr_textmode->VSyncEnd, curr_textmode->VTotal);
    set_H_timings(curr_textmode->HDisplay, curr_textmode->HSyncStart, curr_textmode->HSyncEnd, curr_textmode->HTotal);
    
    interlace(chipset, curr_textmode);
    
    Outbit_CRTC(0x09, 7, MOFLG_ISSET(curr_textmode, ATTR_DOUBLESCAN) ? 1 : 0);

    Set_CURSOR_START(cursor_start) ; Set_CURSOR_END(cursor_end);
    
    if (underline_pos>=0) Set_Underline_location(underline_pos);

    Set_HSYNC_POLARITY(curr_textmode->hpol) ; Set_VSYNC_POLARITY(curr_textmode->vpol);

    Set_HORIZ_SHIFT(curr_textmode->hshift);

    /* set sync-on-green on special RAMDAC's */
    if ( (clock_data.clockchiptype==CLKCHIP_TI3026) || (clock_data.clockchiptype==CLKCHIP_TI3025) )
      set_ti_SOG(chipset, (OFLG_ISSET(OPT_SOG)) );
      
    Outbit_ATR_CTL(0x10, 3, OFLG_ISSET(OPT_16COLOR) ? 0 : 1);

#ifndef DOS
    /* set console to text mode */
    ioctl(opentty("/dev/console"), KDSETMODE, KD_TEXT);
#endif

    if (set_charwidth(curr_textmode->FontWidth)) PERROR(("Illegal character width: %d\n",curr_textmode->FontWidth));

    Outb_ATR_CTL(0x11, bordercolor);
    
    SCREEN_ON;
    
    /* could do a safe_wait_vsync() here to check if we have some refresh */

  }
  
 /*
  * call the external font loading program, if enabled by the 'option loadfont' line
  */

 if ((program_hardware) && (chipset==CS_S3))
 {
   /* S3 high speed text mode cannot be run when font loading is not enabled */
   S3_StartHSText_FontLoad(curr_textmode->pixelClock, FALSE);
 }
 
 if (OFLG_ISSET(OPT_LOADFONT) && (program_hardware) && (run_fontprog))
 {
   if (!(font_data.font_table[curr_textmode->FontWidth-8][curr_textmode->FontHeight-1]))
   {
     PWARNING(("Font loading enabled, but no font specified for %dx%d text mode.\n", curr_textmode->FontWidth, curr_textmode->FontHeight));
     PWARNING(("No font will be loaded.\n"));
   }
   else
   {
     char tempstr[1024]="";
     int result=0;
     snprintf(tempstr, sizeof(tempstr), "%s ", font_data.fontprogpath);
     if (font_data.fontpath)
     {
       strcat(tempstr, font_data.fontpath);
       if (LAST_CHAR(tempstr) != DIR_SEP)
         strcat(tempstr, DIR_SEP_S);
     }
     strcat(tempstr, font_data.font_table[curr_textmode->FontWidth-8][curr_textmode->FontHeight-1]);

     if (chipset==CS_S3)
     {
       PDEBUG(("Executing external command %s'\n", tempstr));
       S3_StartHSText_FontLoad(curr_textmode->pixelClock, TRUE);
       result = Run_extern_Prog_pipe(tempstr);
       S3_EndHSText_FontLoad();
       show_extout();
       if (result<0) PWARNING(("External program call `%s' through pipe failed.\n", tempstr));
     }
     else
     {
       Run_extern_Prog(tempstr);
     }
   }
 }

// Hacking the hack; ResetProg might be useful in DOS, too. - Wolfram, Jan 06
#ifdef DOS
#  undef NO_RESIZE
#endif

#ifndef NO_RESIZE
 /*~
  * call the reset program (which could be used to notify applications about the screen changes)
  * But only when the screen has been resized.
  *
  * Of course, nobody needs this when resizing is impossible (NO_RESIZE defined)
  */

 if ((resetprogpath) && (program_hardware) && (run_resetprog))
 {
    char tempstr[1024]="";
    if (!sresize && !alwaysresize)
      PDEBUG(("Screen not resized. ResetProg not executed.\n"));
    else
    {
      snprintf(tempstr, sizeof(tempstr), "%s %d %d", resetprogpath, curr_textmode->cols, curr_textmode->rows);
      Run_extern_Prog(tempstr);
    }
  }       
#endif

#ifdef DOS
#  define NO_RESIZE
#endif
  return(0);
}

